sdk/communication/Azure.Communication.Administration/src/Models/PhoneNumberReservationOperation.cs

```diff
@@ -0,0 +1,210 @@
+
+        /// <summary>
+        /// Check for the latest status of the operation.
+        /// </summary>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <param name="async" />
+        /// <returns>The <see cref="Response"/> with the status update.</returns>
+        private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
+        {
+            using DiagnosticScope scope = _client.ClientDiagnostics.CreateScope($"{nameof(PhoneNumberReservationOperation)}.{nameof(UpdateStatus)}");
+            scope.Start();
+
+            try
+            {
+                if (HasCompleted)
+                {
+                    return GetRawResponse();
+                }
+
+                if (cancellationToken == default)
+                {
+                    cancellationToken = _cancellationToken;
+                }
+
+                Response<PhoneNumberReservation> update = async
+                    ? await _client.GetReservationByIdAsync(reservationId: Id, cancellationToken: cancellationToken).ConfigureAwait(false)
+                    : _client.GetReservationById(reservationId: Id, cancellationToken: cancellationToken);
+
+                if (!HasCompleted)
+                    _rawResponse = update.GetRawResponse();
+
+                if (IsOperationCompleted(update))
+                {
+                    _value = update.Value;
+                    _hasCompleted = true;
+                }
+
+                return GetRawResponse();
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        private bool IsOperationCompleted(Response<PhoneNumberReservation> response)
+        {
+            if (response.Value?.Status == null)
+                return false;
+
+            return _terminateStatuses.Contains(response.Value.Status.Value);
+        }
+
+        /// <inheritdocs />
+        public override ValueTask<Response<PhoneNumberReservation>> WaitForCompletionAsync(CancellationToken cancellationToken = default)
+            => this.DefaultWaitForCompletionAsync(cancellationToken);
+
+        /// <inheritdocs />
+        public override ValueTask<Response<PhoneNumberReservation>> WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken)
+            => this.DefaultWaitForCompletionAsync(pollingInterval, cancellationToken);
+
+        private static string GetErrorMessage(PhoneNumberReservation? reservation)
+        {
+            var reservationStatus = reservation?.Status;
+
+            if (reservationStatus == ReservationStatus.Cancelled)
+                return "Reservation is cancelled.";
+            if (reservationStatus == ReservationStatus.Error)
+                return "Reservation error.";
```

Comment:
Maybe to be more consistent with the other two messages:

```suggestion
                return "Reservation has failed.";
```